src/inmanta/data/__init__.py

```diff
@@ -2619,13 +2619,121 @@ async def get_compile_reports(
             )
             for compile in compile_records
         ]
         return dtos
 
+    @classmethod
+    async def get_compile_details(cls, environment: uuid.UUID, id: uuid.UUID) -> Optional[m.CompileDetails]:
+        """Find all of the details of a compile, with reports from a substituted compile, if there was one"""
+        query = f"""
+            WITH RECURSIVE compiledetails AS (
+            SELECT
+                c.id,
+                c.remote_id,
+                c.environment,
+                c.requested,
+                c.started,
+                c.completed,
+                c.success,
+                c.version,
+                c.do_export,
+                c.force_update,
+                c.metadata,
+                c.environment_variables,
+                c.compile_data,
+                c.substitute_compile_id,
+                r.id as report_id,
+                r.started report_started,
+                r.completed report_completed,
+                r.command,
+                r.name,
+                r.errstream,
+                r.outstream,
+                r.returncode
+            FROM
+                {cls.table_name()} c LEFT JOIN public.report r on c.id = r.compile
+            WHERE
+                c.environment = $1 AND c.id = $2
+            UNION
+                SELECT
+                    comp.id,
+                    comp.remote_id,
+                    comp.environment,
+                    comp.requested,
+                    comp.started,
+                    comp.completed,
+                    comp.success,
+                    comp.version,
+                    comp.do_export,
+                    comp.force_update,
+                    comp.metadata,
+                    comp.environment_variables,
+                    comp.compile_data,
+                    comp.substitute_compile_id,
+                    rep.id as report_id,
+                    rep.started as report_started,
+                    rep.completed as report_completed,
+                    rep.command,
+                    rep.name,
+                    rep.errstream,
+                    rep.outstream,
+                    rep.returncode
+                FROM
+                    /* Lookup the compile with the id that matches the subsitute_compile_id of the current one */
+                    {cls.table_name()} comp
+                    INNER JOIN compiledetails cd ON cd.substitute_compile_id = comp.id
+                    LEFT JOIN public.report rep on comp.id = rep.compile
+        ) SELECT * FROM compiledetails;
+        """
+        values = [cls._get_value(environment), cls._get_value(id)]
+        result = await cls.select_query(query, values, no_obj=True)
+        result = cast(List[Record], result)
+
+        if not result:
+            return None
+        # Get the details from the requested compile, and the reports from the substitute
+        records = list(filter(lambda r: r["id"] == id, result))
```

Comment:
This I don't understand. I think we need to increase the verbosity of the documentation here.

Reply:
I added some more comments, let me know if you think it should be more detailed